Entering a To: address moves the viewpoint to the bottom right corner of the email

rr-codes opened the pull request:
#### c5a7221120cea780897f84fff076b7d98b818b5a
<pre>
Entering a To: address moves the viewpoint to the bottom right corner of the email
<a href="https://bugs.webkit.org/show_bug.cgi?id=245297">https://bugs.webkit.org/show_bug.cgi?id=245297</a>
rdar://96703879

Reviewed by Tim Horton.

On iPad in the Mail app, when composing a message whose contents are taller
than the viewport size, and when the compose window is narrow enough such that
the contact picker appears as a modal when entering a contact in the `To:`
field, dismissing the contact picker causes the message to erroneously scroll
to the bottom right of the message.

This is because in this case, there is a large top inset on the web view&apos;s
scroll view. This causes the UIProcess to erroneously create a `CGRectNull`
and sends it to the web process in a visible content rect update. Because a
`CGRectNull` has an origin of `(INT_MAX, INT_MAX)`, the web view attempts to
scroll to that position.

This PR fixes this by using `FloatRect`s intersection behavior instead of using
`CGRectIntersection`, so that the resulting rect&apos;s origin stays the same instead
of becoming `CGRectNull` in the case the two rects don&apos;t intersect.

A change also had to be made to
`-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]` so that the
view scrolls to its original position with the inset, instead of `0`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]):
(-[WKWebView _updateVisibleContentRects]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreScrollPosition.mm:
(-[RestoreScrollPositionWithLargeContentInsetWebView safeAreaInsets]):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/254640@main">https://commits.webkit.org/254640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dad9efb47f0b3ed7fbfe9ce4f42642b61764cb3b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34273 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99061 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155879 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32767 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93418 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26041 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76568 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/25994 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/80916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/80784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/68985 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30520 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14868 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30269 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/15805 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3257 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/sandbox-inherit-to-blank-document-unsandboxed.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33719 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/38748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1378 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34893 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreScrollPosition.mm:
```diff
@@ -32,10 +32,45 @@
 #import <WebKit/WKWebViewPrivate.h>
 #import <WebKit/_WKProcessPoolConfiguration.h>
 
+@interface RestoreScrollPositionWithLargeContentInsetWebView: TestWKWebView
```
Space on both sides of the :

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreScrollPosition.mm:
```diff
@@ -32,10 +32,45 @@
+    auto webView = adoptNS([[RestoreScrollPositionWithLargeContentInsetWebView alloc] initWithFrame:CGRectMake(0, 0, 375, 1024)]);
+    
+    [webView synchronouslyLoadTestPageNamed:@"simple-tall"];
+    [[webView scrollView] addSubview:[[UIView alloc] initWithFrame:CGRectMake(0, 0, 375, 600)]];
```
Probably don't actually need this? The space is there even if you don't fill it

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2481,7 +2487,10 @@ - (void)_updateVisibleContentRects
     CGRect unobscuredRect = UIEdgeInsetsInsetRect(self.bounds, computedContentInsetUnadjustedForKeyboard);
     CGRect unobscuredRectInContentCoordinates = _perProcessState.frozenUnobscuredContentRect ? _perProcessState.frozenUnobscuredContentRect.value() : [self convertRect:unobscuredRect toView:_contentView.get()];
+    CGPoint unobscuredRectInContentCoordinatesOrigin = unobscuredRectInContentCoordinates.origin;
     unobscuredRectInContentCoordinates = CGRectIntersection(unobscuredRectInContentCoordinates, [self _contentBoundsExtendedForRubberbandingWithScale:scaleFactor]);
```
Does FloatRect's intersection behavior do what we want? If so, you could consider using that instead of CGRectIntersection and this little dance

rr-codes replied:
@hortont424 what do you mean exactly? The rects being intersected are `CGRect`s, and they can only be intersected one way as far as I know?

hortont424 replied:
FloatRect has:
    WEBCORE_EXPORT FloatRect(const CGRect&);
    WEBCORE_EXPORT operator CGRect() const;
so there's nothing stopping you from using it :)

hortont424 replied:
(but I don't know it's actually saner! you should check)

rr-codes replied:
I mean like what benefit does it have over just using the CGRect intersection like I am currently?

hortont424 replied:
It looks like it's not good enough -- it collapses to {0, 0, 0, 0}, not {x, y, 0, 0}

rr-codes replied:
so do you mean like
```
    WebCore::FloatRect unobscuredRectInContentCoordinates = WebCore::FloatRect(_perProcessState.frozenUnobscuredContentRect ? _perProcessState.frozenUnobscuredContentRect.value() : [self convertRect:unobscuredRect toView:_contentView.get()]);
    unobscuredRectInContentCoordinates.intersect([self _contentBoundsExtendedForRubberbandingWithScale:scaleFactor]);
    ```

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1287,9 +1287,15 @@ - (void)_scrollToContentScrollPosition:(WebCore::FloatPoint)scrollPosition scrol
     [_scrollView _stopScrollingAndZoomingAnimations];
 
-    if (!CGPointEqualToPoint(contentOffsetInScrollViewCoordinates, [_scrollView contentOffset]))
+    if (!CGPointEqualToPoint(contentOffsetInScrollViewCoordinates, [_scrollView contentOffset])) {
```
Maybe store `[_scrollView contentOffset]` in a local variable? It's used five times here.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2480,8 +2488,8 @@ - (void)_updateVisibleContentRects
     CGRect unobscuredRect = UIEdgeInsetsInsetRect(self.bounds, computedContentInsetUnadjustedForKeyboard);
-    CGRect unobscuredRectInContentCoordinates = _perProcessState.frozenUnobscuredContentRect ? _perProcessState.frozenUnobscuredContentRect.value() : [self convertRect:unobscuredRect toView:_contentView.get()];
-    unobscuredRectInContentCoordinates = CGRectIntersection(unobscuredRectInContentCoordinates, [self _contentBoundsExtendedForRubberbandingWithScale:scaleFactor]);
+    WebCore::FloatRect unobscuredRectInContentCoordinates = WebCore::FloatRect(_perProcessState.frozenUnobscuredContentRect ? _perProcessState.frozenUnobscuredContentRect.value() : [self convertRect:unobscuredRect toView:_contentView.get()]);
+    unobscuredRectInContentCoordinates.intersect([self _contentBoundsExtendedForRubberbandingWithScale:scaleFactor]);
```
Is this correct? The PR description says:

```
This PR fixes this by checking if the CGRect is `CGRectNull`, and if so,
replaces its origin with the original unobscured rect's origin.
```

Does `intersect` preserve the original origin? If not, wouldn't this cause an unexpected scroll to (0, 0)?

rr-codes replied:
Yes, the comment was before I applied Tim's suggestion. It does preserve the original origin; I tested it manually.

rr-codes committed:
Entering a To: address moves the viewpoint to the bottom right corner of the email
https://bugs.webkit.org/show_bug.cgi?id=245297
rdar://96703879

Reviewed by Tim Horton.

On iPad in the Mail app, when composing a message whose contents are taller
than the viewport size, and when the compose window is narrow enough such that
the contact picker appears as a modal when entering a contact in the `To:`
field, dismissing the contact picker causes the message to erroneously scroll
to the bottom right of the message.

This is because in this case, there is a large top inset on the web view's
scroll view. This causes the UIProcess to erroneously create a `CGRectNull`
and sends it to the web process in a visible content rect update. Because a
`CGRectNull` has an origin of `(INT_MAX, INT_MAX)`, the web view attempts to
scroll to that position.

This PR fixes this by using `FloatRect`s intersection behavior instead of using
`CGRectIntersection`, so that the resulting rect's origin stays the same instead
of becoming `CGRectNull` in the case the two rects don't intersect.

A change also had to be made to
`-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]` so that the
view scrolls to its original position with the inset, instead of `0`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]):
(-[WKWebView _updateVisibleContentRects]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreScrollPosition.mm:
(-[RestoreScrollPositionWithLargeContentInsetWebView safeAreaInsets]):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/254640@main
